interfaces/builtin/serial_port.go

```diff
@@ -50,37 +59,83 @@ func (iface *SerialPortInterface) SanitizeSlot(slot *interfaces.Slot) error {
 		panic(fmt.Sprintf("slot is not of interface %q", iface))
 	}
 
+	// Allow the core snap to create implicits serial-port slots with no
+	// attributes. These will be used to grant access based on Udev rules
+	if slot.Snap.Type == "os" && len(slot.Attrs) == 0 {
+		return nil
+	}
+
+	if len(slot.Attrs) > 1 {
+		return fmt.Errorf("serial-port slot definition has unexpected number of attributes")
+	}
+
 	// Check slot has a path attribute identify serial device
 	path, ok := slot.Attrs["path"].(string)
 	if !ok || path == "" {
-		return fmt.Errorf("serial-port slot must have a path attribute")
+		return fmt.Errorf(`serial-port slot does not have "path" attribute`)
 	}
 
 	// Clean the path before checking it matches the pattern
 	path = filepath.Clean(path)
 
 	// Check the path attribute is in the allowable pattern
-	if serialAllowedPathPattern.MatchString(path) {
-		return nil
+	if !serialAllowedPathPattern.MatchString(path) {
+		return fmt.Errorf("serial-port path attribute must be a valid device node")
 	}
-	return fmt.Errorf("serial-port path attribute must be a valid device node")
+
+	return nil
 }
 
 // SanitizePlug checks and possibly modifies a plug.
-func (iface *SerialPortInterface) SanitizePlug(slot *interfaces.Plug) error {
-	if iface.Name() != slot.Interface {
+func (iface *SerialPortInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
 		panic(fmt.Sprintf("plug is not of interface %q", iface))
 	}
-	// NOTE: currently we don't check anything on the plug side.
+
+	switch len(plug.Attrs) {
+	case 1:
+		// In the case of one attribute it should be valid path attribute
+		// Check slot has a path attribute identify serial device
```

Comment:
Ok, now that we are on the plugs side...

This comment has a grammatical error. I think it is supposed to say "Check slot has a path attribute *to* identify serial device".

That said, the code doesn't do that. It seems to me that the case 1 code needs to:
- if has path attribute, verify the pattern and check if provided by the slot side
- if has vendor attribute, verify the pattern
- if has product attribute, verify the pattern